R/reporter-junit.R

```diff
@@ -0,0 +1,131 @@
+#' @include reporter.R
+NULL
+
+#' Test reporter: summary of errors in jUnit XML format.
+#'
+#' This reporter includes detailed results about each test and summaries,
+#' written to a file (or stdout) in jUnit XML format. This can be read by
+#' the Jenkins Continuous Integration System to report on a dashboard etc.
+#' Requires the XML package.
+#'
+#' This works for \code{\link{expect_that}} but not for the wrappers like
+#' \code{\link{expect_equal}} etc.
+#'
+#' To fit into the jUnit structure, context() becomes the \code{<testsuite>}
+#' name as well as the base of the \code{<testcase> classname}. The
+#' test_that() name becomes the rest of the \code{<testcase> classname}.
+#' The deparsed expect_that() call becomes the \code{<testcase>} name.
+#' On failure, the message goes into the \code{<failure>} node message
+#' argument (first line only) and into its text content (full message).
+#'
+#' Execution time and some other details are also recorded.
+#'
+#' References for the jUnit XML format:
+#' \url{http://llg.cubic.org/docs/junit/}
+#'
+#' Output drawn from the SummaryReporter is printed to the standard
+#' error stream during execution.
+#'
+#' @export
+JunitReporter <- R6::R6Class("JunitReporter", inherit = Reporter,
+  public = list(
+    file = NULL,
+    results = NULL,
+    timer = NULL,
+
+    initialize = function(file = "report.xml") {
+      super$initialize()
+      if (!require("XML", quietly = TRUE)) {
+        stop("Please install the XML package", call. = FALSE)
+      }
+      self$file <- file
+      self$results <- list()
+    },
+
+    start_reporter = function() {
+      self$timer <- proc.time()
+    },
+
+    start_context = function(context) {
+      self$cat(context, ": ")
+    },
+
+    end_context = function(context) {
+      self$cat("\n")
+    },
+
+    add_result = function(context, test, result) {
+      if (expectation_broken(result)) {
+        self$cat_tight(single_letter_summary(result))
+      }else {
+        self$cat_tight(colourise(".", "passed"))
+      }
+
+      result$time <- round((proc.time() - self$timer)[["elapsed"]], 2)
+      self$timer  <- proc.time()
+      result$test <- if (is.null(test) || test == "") "(unnamed)" else test
+      # call can sometimes contain a second item, "succeed()"
+      result$call <- if (is.null(result$call)) "(unexpected)" else format(result$call)[1]
+      self$results[[context]] <- append(self$results[[context]], list(result))
+    },
+
+    end_reporter = function() {
+      self$cat("\n")
+      xmlNodeOK <- function(name, ..., attrs = NULL) {
+        ## do XML entity substitutions
+        if (!is.null(attrs)) {
+          attrs <- vapply(attrs, function(x) toString(xmlTextNode(x)), character(1))
+        }
+        xmlNode(name, ..., attrs = attrs)
+      }
+      classnameOK <- function(text) {
+        gsub("[ \\.]", "_", text)
+      }
+      suites <- lapply(names(self$results), function(context) {
```

Comment:
I think this would be easier to follow if you pulled out separate functions for the expectation, test, and context levels